<a href="https://colab.research.google.com/github/JaydaBubel/lab-random-forests/blob/master/lab_random_forests.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [50]:
import pandas as pd
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import confusion_matrix, classification_report

In [51]:
df_numerical=pd.read_csv("/content/drive/MyDrive/categorical.csv")
df_categorical=pd.read_csv("/content/drive/MyDrive/numerical.csv")
df_target=pd.read_csv("/content/drive/MyDrive/target.csv")

In [52]:
df_categorical

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
0          0  60.000000       5        9    0         0       39        34   
1          1  46.000000       6        9   16         0       15        55   
2          1  61.611649       3        1    2         0       20        29   
3          0  70.000000       1        4    2         0       23        14   
4          0  78.000000       3        2   60         1       28         9   
...      ...        ...     ...      ...  ...       ...      ...       ...   
95407      1  61.611649       5        9    0        14       36        47   
95408      1  48.000000       7        9    1         0       31        43   
95409      1  60.000000       5        9    0         0       18        46   
95410      0  58.000000       7        9    0         0       28        35   
95411      2  80.000000       5        8    3         1       33        22   

       WWIIVETS  LOCALGOV  ...  CARDGIFT  MINRAMNT  MAXRAMNT  LASTGIFT  \
0            18        10  ...        14       5.0      12.0      10.0   
1            11         6  ...         1      10.0      25.0      25.0   
2            33         6  ...        14       2.0      16.0       5.0   
3            31         3  ...         7       2.0      11.0      10.0   
4            53        26  ...         8       3.0      15.0      15.0   
...         ...       ...  ...       ...       ...       ...       ...   
95407        11         7  ...         0      25.0      25.0      25.0   
95408        19         4  ...         0      20.0      20.0      20.0   
95409        20         7  ...         4       3.0      10.0      10.0   
95410        20         9  ...        18       5.0      21.0      18.0   
95411        27         8  ...        11      10.0     201.0     100.0   

       TIMELAG    AVGGIFT  CONTROLN  HPHONE_D  RFA_2F  CLUSTER2  
0            4   7.741935     95515         0       4        39  
1           18  15.666667    148535         0       2         1  
2           12   7.481481     15078         1       4        60  
3            9   6.812500    172556         1       4        41  
4           14   6.864865      7112         1       2        26  
...        ...        ...       ...       ...     ...       ...  
95407        9  25.000000    184568         0       1        12  
95408        9  20.000000    122706         1       1         2  
95409        3   8.285714    189641         1       3        34  
95410        4  12.146341      4693         1       4        11  
95411        6  96.794872    185114         1       1        12  

[95412 rows x 315 columns]

In [53]:
df_target

TARGET_B  TARGET_D
0             0       0.0
1             0       0.0
2             0       0.0
3             0       0.0
4             0       0.0
...         ...       ...
95407         0       0.0
95408         0       0.0
95409         0       0.0
95410         1      18.0
95411         0       0.0

[95412 rows x 2 columns]

In [54]:
column_names = df_categorical.columns.tolist()
print("Column Names:", column_names)

Column Names: ['TCODE', 'AGE', 'INCOME', 'WEALTH1', 'HIT', 'MALEMILI', 'MALEVET', 'VIETVETS', 'WWIIVETS', 'LOCALGOV', 'STATEGOV', 'FEDGOV', 'WEALTH2', 'POP901', 'POP902', 'POP903', 'POP90C1', 'POP90C2', 'POP90C3', 'POP90C4', 'POP90C5', 'ETH1', 'ETH2', 'ETH3', 'ETH4', 'ETH5', 'ETH6', 'ETH7', 'ETH8', 'ETH9', 'ETH10', 'ETH11', 'ETH12', 'ETH13', 'ETH14', 'ETH15', 'ETH16', 'AGE901', 'AGE902', 'AGE903', 'AGE904', 'AGE905', 'AGE906', 'AGE907', 'CHIL1', 'CHIL2', 'CHIL3', 'AGEC1', 'AGEC2', 'AGEC3', 'AGEC4', 'AGEC5', 'AGEC6', 'AGEC7', 'CHILC1', 'CHILC2', 'CHILC3', 'CHILC4', 'CHILC5', 'HHAGE1', 'HHAGE2', 'HHAGE3', 'HHN1', 'HHN2', 'HHN3', 'HHN4', 'HHN5', 'HHN6', 'MARR1', 'MARR2', 'MARR3', 'MARR4', 'HHP1', 'HHP2', 'DW1', 'DW2', 'DW3', 'DW4', 'DW5', 'DW6', 'DW7', 'DW8', 'DW9', 'HV1', 'HV2', 'HV3', 'HV4', 'HU1', 'HU2', 'HU3', 'HU4', 'HU5', 'HHD1', 'HHD2', 'HHD3', 'HHD4', 'HHD5', 'HHD6', 'HHD7', 'HHD8', 'HHD9', 'HHD10', 'HHD11', 'HHD12', 'ETHC1', 'ETHC2', 'ETHC3', 'ETHC4', 'ETHC5', 'ETHC6', 'HVP1', 'H

In [55]:
categorical_columns = ["STATE","HOMEOWNR","GENDER","RFA_2R","RFA_2A","GEOCODE2","DOMAIN_A"]

In [56]:
df_target['TARGET_B'].value_counts()

0    90569
1     4843
Name: TARGET_B, dtype: int64

In [57]:
df_target['TARGET_D'].value_counts()

0.00     90569
10.00      941
15.00      591
20.00      577
5.00       503
         ...  
18.25        1
10.70        1
2.50         1
16.87        1
44.21        1
Name: TARGET_D, Length: 71, dtype: int64

In [58]:
df_target = df_target.drop('TARGET_D', axis = 1)

In [59]:
df = pd.concat([df_categorical, df_numerical], axis=1)

df.head()

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
0      0  60.000000       5        9    0         0       39        34   
1      1  46.000000       6        9   16         0       15        55   
2      1  61.611649       3        1    2         0       20        29   
3      0  70.000000       1        4    2         0       23        14   
4      0  78.000000       3        2   60         1       28         9   

   WWIIVETS  LOCALGOV  ...  DOB_YR  DOB_MM  MINRDATE_YR  MINRDATE_MM  \
0        18        10  ...      37      12           92            8   
1        11         6  ...      52       2           93           10   
2        33         6  ...       0       2           91           11   
3        31         3  ...      28       1           87           11   
4        53        26  ...      20       1           93           10   

   MAXRDATE_YR  MAXRDATE_MM  LASTDATE_YR  LASTDATE_MM  FIRSTDATE_YR  \
0           94            2           95           12            89   
1           95           12           95           12            93   
2           92            7           95           12            90   
3           94           11           95           12            87   
4           96            1           96            1            79   

   FIRSTDATE_MM  
0            11  
1            10  
2             1  
3             2  
4             3  

[5 rows x 337 columns]

In [60]:
X = pd.get_dummies(df)
y = df_target['TARGET_B']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

smote = SMOTE()
X_train, y_train = smote.fit_resample(X_train, y_train)

classification = RandomForestClassifier(max_depth=2,random_state=0)
classification.fit(X_train, y_train)
predictions = classification.predict(X_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.95      0.79      0.86     18114
           1       0.07      0.29      0.11       969

    accuracy                           0.76     19083
   macro avg       0.51      0.54      0.49     19083
weighted avg       0.91      0.76      0.83     19083



Precision: for class 0, model is 95% accurate in predictions, but for class 1, only 6% accurate -- it's much better at identifying class 0.

Recall: for class 0, model correctly finds 79% of all class 0 cases. For class 1, it finds 27% of the class 1 cases -- it's not good at identifiyng class 1 cases

Accuracy: model is 77% accurate, but biased towards class 0.
Baisclaly, the model is good at getting class 0 right but not class 1.
